cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java

```diff
@@ -0,0 +1,211 @@
+
+    if (itc == null) {
+      throw new IllegalStateException("FileNamePattern ["
+                                        + fileNamePattern.getPattern()
+                                        + "] does not contain a valid IntegerToken");
+    }
+    processedIndex = maxIndex;
+    super.start();
+  }
+
+  @Override
+  public void rollover() throws RolloverFailure {
+    Location parentLocation = getParent(activeFileLocation);
+
+    if (parentLocation == null) {
+      return;
+    }
+
+    // If maxIndex <= 0, then there is no file renaming to be done.
+    if (maxIndex >= 0) {
+      try {
+        String fileName = fileNamePattern.convertInt(maxIndex);
+        // Delete the oldest file. If processedIndex is not pointing to maxIndex, that means there was some
+        // exception while doing rename of files in previous rollover. So do not delete the existing file with
+        // maximum index otherwise we will end up deleting rolled over file
+        if (processedIndex == maxIndex) {
+          Location deleteLocation = parentLocation.append(fileName);
+          // no need to proceed further if we are not able to delete location so throw exception
+          if (deleteLocation.exists() && !deleteLocation.delete()) {
+            LOG.warn("Failed to delete location: {}", deleteLocation.toURI().toString());
+            throw new RolloverFailure(String.format("Not able to delete file: %s", deleteLocation.toURI().toString()));
+          }
+          processedIndex--;
+        }
+
+        for (int i = processedIndex; i >= minIndex; i--) {
+          String toRenameStr = fileNamePattern.convertInt(i);
+          Location toRename = parentLocation.append(toRenameStr);
+
+          // no point in trying to rename an non existent file
+          if (toRename.exists()) {
+            Location newName = parentLocation.append(fileNamePattern.convertInt(i + 1));
+            // throw exception if rename fails, so that in next iteration of rollover, it will be retried
+            if (toRename.renameTo(newName) == null) {
+              processedIndex = i;
+              LOG.warn("Failed to rename {} to {}", toRename.toURI().toString(), newName.toURI().toString());
+              throw new RolloverFailure(String.format("Failed to rename %s to %s", toRename.toURI().toString(),
+                                                      newName.toURI().toString()));
+            }
+          } else {
+            LOG.trace("Skipping roll-over for inexistent file {}", toRename.toURI().toString());
+          }
+        }
+
+        // we need to set processedIndex to lower than minIndex because if any failure happens after this point, we
+        // do not want to reprocess all the files we just renamed
+        processedIndex = minIndex - 1;
+
+        // close outputstream of active location before renaming it
+        closeable.close();
+
+        if (activeFileLocation.renameTo(parentLocation.append(fileNamePattern.convertInt(minIndex))) == null) {
```

Comment:
what happens when this rename fails ? don't think that's handled for retry. as closeable's close succeeding would have removed this from the active locations map, right ?